thewalrus/tests/test_decompositions.py

```diff
@@ -147,3 +149,33 @@ def test_mixed_state(self, create_cov, hbar, tol):
         assert np.allclose(sorted(nbar[:n]), sorted(nbar_in), atol=tol, rtol=0)
         # check S is symplectic
         assert np.allclose(S @ O @ S.T, O, atol=tol, rtol=0)
+
+
+class TestBlochMessiahDecomposition:
+    """Tests for the Williamson decomposition"""
+
+    @pytest.mark.parametrize("N", range(50, 500, 50))
+    def test_blochmessiah_rand(self, N):
+        """Tests blochmessiah function for different matrix sizes."""
+        S = random_symplectic(N)
+        u, d, v = blochmessiah(S)
+        assert np.allclose(u @ d @ v, S)
+        np.allclose(u.T @ u, np.eye(len(u)))
+        np.allclose(v.T @ v, np.eye(len(v)))
+        is_symplectic(u)
+        is_symplectic(v)
+
+    def test_blochmessiah_odd(self):
+        """Tests that odd matrices return False in blochmessiah."""
+        S = np.random.rand(5, 5)
+        assert not blochmessiah(S)
+
+    def test_blochmessiah_rect(self):
+        """Tests that rectangular matrices return False in blochmessiah"""
+        S = np.random.rand(4, 5)
+        assert not blochmessiah(S)
+
+    def test_blochmessiah_false(self):
+        """Tests that non-symplectic mattrices return False in blochmessiah"""
+        S = np.random.rand(4, 4)
+        assert not blochmessiah(S)
```

Comment:
Given the suggestion of raising a `ValueError` is implemented, this test should be refactored accordingly.